src/Gui/ExpressionCompleter.cpp

```diff
@@ -144,14 +231,32 @@ class ExpressionCompleterModel: public QAbstractItemModel {
         QVariant v;
         Info info = getInfo(index);
         _data(info,index.row(),&v,nullptr,role==Qt::UserRole);
-        FC_TRACE(info.d.doc << "," << info.d.obj << "," << index.row()
+        FC_TRACE(  info.doc << "," << info.obj << "," << info.prop << "," 
+                << info.contextualHierarchy << "," << index.row()
                 << ": " << v.toString().toUtf8().constData());
         return v;
     }
 
+   // The completion tree structure created takes into account the current document and object
+    //
+    // It is done as such:
+    // * root (-1,-1)
+    // |
```

Comment:
This order causes issues with contextual completion. 
If you have a document, or an object with the same name as a property, you will not be able to use completion to reach that property, because the completer's contextual solve will pick up the object rather than the prop.

Even swapping these around will still make it ambiguous.  (did you mean the prop or the object)

The Assembly 4 workbench has an issue that the model contains an 'Constraints' Object :(. As such any local object referencing "Constraints" won't be resolved properly.

Reply:
It may work to swap these around, it depends if the split path code parser doesn't collapse the current document references.

Maybe there is a workaround (with ._self)? but I don't know FreeCAD very well.